File: python/plan/astar3d.py
import heapq
import math
from typing import List, Tuple, Dict, Set
from map3d import Map3D

class Node:
    """
    A*算法中的节点类
    """
    def __init__(self, x: float, y: float, z: float, g: float = 0, h: float = 0, parent = None):
        self.x = x
        self.y = y
        self.z = z
        self.g = g  # 从起点到当前节点的实际代价
        self.h = h  # 启发式函数估算的从当前节点到终点的代价
        self.f = g + h  # 总代价
        self.parent = parent  # 父节点
        
    def __lt__(self, other):
        # 优化比较操作，首先比较f值，然后比较h值（更接近目标的优先）
        if self.f == other.f:
            return self.h < other.h
        return self.f < other.f
        
    def __eq__(self, other):
        return isinstance(other, Node) and self.x == other.x and self.y == other.y and self.z == other.z
        
    def __hash__(self):
        # 使用坐标哈希，但需要处理浮点数精度问题
        return hash((round(self.x, 6), round(self.y, 6), round(self.z, 6)))

class AStar3D:
    """
    3D A*路径规划算法
    """
    
    def __init__(self, map3d: Map3D):
        self.map = map3d
        
    def heuristic(self, x1: float, y1: float, z1: float, x2: float, y2: float, z2: float) -> float:
        """
        启发式函数，计算两个点之间的估计代价
        
        Args:
            x1, y1, z1: 起点世界坐标
            x2, y2, z2: 终点世界坐标
            
        Returns:
            float: 估计代价
        """
        # 使用欧几里得距离作为启发式函数
        return math.sqrt((x1 - x2)**2 + (y1 - y2)**2 + (z1 - z2)**2)
        
    def _get_key(self, x: float, y: float, z: float) -> Tuple[int, int, int]:
        """
        将世界坐标转换为网格坐标作为键值，用于快速查找
        
        Args:
            x, y, z: 世界坐标
            
        Returns:
            Tuple[int, int, int]: 网格坐标
        """
        grid_x, grid_y, grid_z = self.map._world_to_grid(x, y, z)
        return (grid_x, grid_y, grid_z)
        
    def plan(self, start: Tuple[float, float, float], goal: Tuple[float, float, float]) -> List[Tuple[float, float, float]]:
        """
        执行A*路径规划
        
        Args:
            start: 起点世界坐标 (x, y, z)
            goal: 终点世界坐标 (x, y, z)
            
        Returns:
            List[Tuple[float, float, float]]: 路径点列表（世界坐标），如果找不到路径则返回空列表
        """
        start_x, start_y, start_z = start
        goal_x, goal_y, goal_z = goal
        
        # 检查起点和终点是否有效
        if not self.map.is_valid_position(start_x, start_y, start_z):
            raise ValueError("Invalid start position")
            
        if not self.map.is_valid_position(goal_x, goal_y, goal_z):
            raise ValueError("Invalid goal position")
            
        if self.map.is_obstacle(start_x, start_y, start_z):
            raise ValueError("Start position is an obstacle")
            
        if self.map.is_obstacle(goal_x, goal_y, goal_z):
            raise ValueError("Goal position is an obstacle")
            
        # 初始化开放列表和关闭列表
        # 使用优先队列存储 (f_score, insertion_index, Node)
        open_list = []
        insertion_index = 0  # 用于保持堆的稳定性
        
        # 使用网格坐标作为键值的字典，提高查找效率
        open_dict: Dict[Tuple[int, int, int], Tuple[float, int, Node]] = {}
        closed_set: Set[Tuple[int, int, int]] = set()
        
        # 创建起点节点
        start_node = Node(start_x, start_y, start_z, 0, self.heuristic(start_x, start_y, start_z, goal_x, goal_y, goal_z))
        heapq.heappush(open_list, (start_node.f, insertion_index, start_node))
        insertion_index += 1
        start_key = self._get_key(start_x, start_y, start_z)
        open_dict[start_key] = (0, 0, start_node)
        
        while open_list:
            # 取出f值最小的节点
            _, _, current_node = heapq.heappop(open_list)
            current_key = self._get_key(current_node.x, current_node.y, current_node.z)
            
            # 检查是否已经在关闭列表中
            if current_key in closed_set:
                continue
            
            # 从开放字典中移除
            if current_key in open_dict:
                del open_dict[current_key]
            
            # 如果到达目标节点（检查是否足够接近目标）
            distance_to_goal = math.sqrt((current_node.x - goal_x)**2 + 
                                       (current_node.y - goal_y)**2 + 
                                       (current_node.z - goal_z)**2)
            if distance_to_goal <= self.map.resolution:
                # 重构路径
                path = []
                while current_node:
                    path.append((current_node.x, current_node.y, current_node.z))
                    current_node = current_node.parent
                return path[::-1]  # 反转路径
                
            # 将当前节点加入关闭列表
            closed_set.add(current_key)
            
            # 检查所有邻居节点
            for neighbor_pos in self.map.get_neighbors(current_node.x, current_node.y, current_node.z):
                # 获取邻居节点的世界坐标
                nx, ny, nz = neighbor_pos
                # 计算邻居的网格键
                neighbor_key = self._get_key(nx, ny, nz)
                
                # 如果在关闭列表中，跳过
                if neighbor_key in closed_set:
                    continue
                
                # 计算到邻居节点的代价
                # 移动代价（欧几里得距离）
                move_cost = math.sqrt((current_node.x - nx)**2 + (current_node.y - ny)**2 + (current_node.z - nz)**2)
                tentative_g_score = current_node.g + move_cost
                
                # 如果我们已经找到了到达这个节点的更好路径，跳过
                if neighbor_key in open_dict and tentative_g_score >= open_dict[neighbor_key][0]:
                    continue
                    
                # 这条路径是到达该节点的最佳路径
                neighbor_node = Node(nx, ny, nz, tentative_g_score, self.heuristic(nx, ny, nz, goal_x, goal_y, goal_z), current_node)
                
                # 添加到开放列表
                heapq.heappush(open_list, (neighbor_node.f, insertion_index, neighbor_node))
                open_dict[neighbor_key] = (tentative_g_score, insertion_index, neighbor_node)
                insertion_index += 1
                        
        # 未找到路径
        return []

File: cpp/plan/AStar3D.cpp
#include "AStar3D.h"
#include <queue>
#include <functional>
#include <cmath>
#include <limits>
#include <algorithm>
#include <set>
#include <map>
#include <unordered_map>
#include <unordered_set>

AStar3D::AStar3D(std::shared_ptr<Map3D> map) : map_(map) {}

std::vector<std::tuple<double, double, double>> AStar3D::plan(
    std::tuple<double, double, double> start,
    std::tuple<double, double, double> goal) {
    
    // No need to convert start and goal to grid coordinates here
    // We'll work directly with world coordinates and use Map3D's conversion methods when needed
    
    // Check if start and goal are valid
    if (!map_->isValidPosition(std::get<0>(start), 
                               std::get<1>(start), 
                               std::get<2>(start)) ||
        !map_->isValidPosition(std::get<0>(goal), 
                               std::get<1>(goal), 
                               std::get<2>(goal)) ||
        map_->isObstacle(std::get<0>(start), 
                         std::get<1>(start), 
                         std::get<2>(start)) ||
        map_->isObstacle(std::get<0>(goal), 
                         std::get<1>(goal), 
                         std::get<2>(goal))) {
        return {}; // Return empty path if start or goal is invalid
    }
    
    // Initialize open and closed sets
    std::priority_queue<Node, std::vector<Node>, CompareNodes> open_set;
    std::unordered_set<Node> closed_set;  // Using custom hash now
    
    // Initialize cost maps
    std::unordered_map<Node, double> g_costs;  // Using custom hash
    std::unordered_map<Node, Node> came_from;  // Using custom hash
    
    // Create start node in world coordinates
    Node start_node(std::get<0>(start), 
                    std::get<1>(start), 
                    std::get<2>(start), 
                    0, 
                    heuristic(std::get<0>(start), 
                              std::get<1>(start), 
                              std::get<2>(start),
                              std::get<0>(goal), 
                              std::get<1>(goal), 
                              std::get<2>(goal)));
    
    open_set.push(start_node);
    g_costs[start_node] = 0;
    
    // Define the 26 neighbors (3D connectivity)
    std::vector<std::tuple<int, int, int>> directions = {
        {-1, -1, -1}, {-1, -1, 0}, {-1, -1, 1},
        {-1, 0, -1},  {-1, 0, 0},  {-1, 0, 1},
        {-1, 1, -1},  {-1, 1, 0},  {-1, 1, 1},
        {0, -1, -1},  {0, -1, 0},  {0, -1, 1},
        {0, 0, -1},              {0, 0, 1},
        {0, 1, -1},   {0, 1, 0},   {0, 1, 1},
        {1, -1, -1},  {1, -1, 0},  {1, -1, 1},
        {1, 0, -1},   {1, 0, 0},   {1, 0, 1},
        {1, 1, -1},   {1, 1, 0},   {1, 1, 1}
    };
    
    while (!open_set.empty()) {
        Node current = open_set.top();
        open_set.pop();
        
        // Check if we reached the goal using tolerance for floating point precision
        const double tolerance = 0.001;
        if (std::abs(current.x - std::get<0>(goal)) < tolerance &&
            std::abs(current.y - std::get<1>(goal)) < tolerance &&
            std::abs(current.z - std::get<2>(goal)) < tolerance) {
            // Reconstruct path
            std::vector<std::tuple<double, double, double>> path;
            Node current_node = current;
            
            // Backtrack to find the complete path
            while(true) {
                path.push_back(std::make_tuple(current_node.x, current_node.y, current_node.z));
                
                // Check if we've reached the start node
                if (std::abs(current_node.x - std::get<0>(start)) < tolerance && 
                    std::abs(current_node.y - std::get<1>(start)) < tolerance && 
                    std::abs(current_node.z - std::get<2>(start)) < tolerance) {
                    break;
                }
                
                // Get the parent of the current node
                if (came_from.find(current_node) != came_from.end()) {
                    current_node = came_from[current_node];
                } else {
                    // If we can't find a parent, we've reached the start
                    break;
                }
            }
            
            std::reverse(path.begin(), path.end());
            return path;
        }
        
        // Add current to closed set
        closed_set.insert(current);
        
        // Explore neighbors
        for (const auto& dir : directions) {
            int new_grid_x = 0, new_grid_y = 0, new_grid_z = 0;
            
            // Calculate new grid coordinates based on current position
            auto [current_grid_x, current_grid_y, current_grid_z] = map_->worldToGrid(current.x, current.y, current.z);
            new_grid_x = current_grid_x + std::get<0>(dir);
            new_grid_y = current_grid_y + std::get<1>(dir);
            new_grid_z = current_grid_z + std::get<2>(dir);
            
            // Check if neighbor is valid and not in closed set
            if (new_grid_x < 0 || new_grid_x >= map_->getWidth() ||
                new_grid_y < 0 || new_grid_y >= map_->getHeight() ||
                new_grid_z < 0 || new_grid_z >= map_->getDepth()) {
                continue;
            }
            
            // Convert back to world coordinates
            auto [nx, ny, nz] = map_->gridToWorld(new_grid_x, new_grid_y, new_grid_z);
            
            Node neighbor_node(nx, ny, nz);
            if (closed_set.count(neighbor_node)) {
                continue;
            }
            
            if (map_->isObstacle(nx, ny, nz)) {
                continue;
            }
            
            // Calculate tentative g_cost
            double dx = nx - current.x;
            double dy = ny - current.y;
            double dz = nz - current.z;
            double move_cost = std::sqrt(dx*dx + dy*dy + dz*dz);
            double tentative_g = current.g + move_cost;
            
            // Check if this path to neighbor is better
            if (g_costs.find(neighbor_node) == g_costs.end() || 
                tentative_g < g_costs[neighbor_node]) {
                
                // Update costs and add to open set
                came_from[neighbor_node] = current;  // Store the current node as parent
                g_costs[neighbor_node] = tentative_g;
                
                Node neighbor(nx, ny, nz,
                              tentative_g,
                              heuristic(nx, ny, nz,
                                        std::get<0>(goal), 
                                        std::get<1>(goal), 
                                        std::get<2>(goal)));
                
                open_set.push(neighbor);
            }
        }
    }
    
    // No path found
    return {};
}

double AStar3D::heuristic(double x1, double y1, double z1, double x2, double y2, double z2) const {
    // Euclidean distance heuristic
    double dx = x2 - x1;
    double dy = y2 - y1;
    double dz = z2 - z1;
    
    return std::sqrt(dx*dx + dy*dy + dz*dz);
}

